pallets/liquidity/src/lib.rs

```diff
@@ -0,0 +1,289 @@
+		) -> DispatchResult {
+			//ensure called by governance
+			T::GovernanceOrigin::ensure_origin(origin)?;
+
+			//ensure account not register already
+			ensure!(
+				!<RegisterGovernanceAccounts<T>>::contains_key(account_generation_key),
+				Error::<T>::PalletAlreadyRegistered
+			);
+
+			//create main account and proxy account
+			let main_account = Self::generate_proxy_account(account_generation_key)?;
+
+			let proxy_account = Self::generate_main_account(account_generation_key)?;
+
+			//call ocex register
+			T::CallOcex::on_register(main_account.clone(), proxy_account.clone())?;
+
+			//insert accounts in storage
+			<RegisterGovernanceAccounts<T>>::insert(
+				account_generation_key,
+				(main_account.clone(), proxy_account.clone()),
+			);
+			Self::deposit_event(Event::PalletAccountRegister { main_account, proxy_account });
+			Ok(())
+		}
+
+		/// Deposit assets to orderbook
+		///
+		/// # Parameters
+		///
+		/// * `origin`: governance
+		/// * `asset`: asset id to deposit
+		/// * `amount`: amount to deposit
+		/// * `account_generation_key`: u32 value that was used to generate main account and proxy
+		///   account
+		#[pallet::weight(10_000)]
+		pub fn deposit_to_orderbook(
+			origin: OriginFor<T>,
+			asset: AssetId,
+			amount: BalanceOf<T>,
+			account_generation_key: u32,
+		) -> DispatchResult {
+			//ensure called by governance
+			T::GovernanceOrigin::ensure_origin(origin)?;
+
+			//check if the account present
+			let (main_account, _) =
+				<RegisterGovernanceAccounts<T>>::try_get(account_generation_key)
+					.map_err(|_| Error::<T>::PalletAccountNotRegistered)?;
+
+			//call ocex deposit
+			T::CallOcex::on_deposit(main_account, asset, amount.saturated_into())?;
+			Ok(())
+		}
+
+		/// Withdraw assets from orderbook
+		///
+		/// # Parameters
+		///
+		/// * `origin`: governance
+		/// * `asset`: asset id to withdraw
+		/// * `amount`: amount to withdraw
+		/// * `do_force_withdraw`: if set to true all active orders will be canceled and then the
+		/// * `account_generation_key`: u32 value that was used to generate main account and proxy
+		///   account given amount will be withdrawn
+		#[pallet::weight(10_000)]
+		pub fn withdraw_from_orderbook(
+			origin: OriginFor<T>,
+			asset: AssetId,
+			amount: BalanceOf<T>,
+			do_force_withdraw: bool,
+			account_generation_key: u32,
+		) -> DispatchResult {
+			T::GovernanceOrigin::ensure_origin(origin)?;
+
+			//check if the account present
+			let (main_account, proxy_account) =
+				<RegisterGovernanceAccounts<T>>::try_get(account_generation_key)
+					.map_err(|_| Error::<T>::PalletAccountNotRegistered)?;
+
+			//call ocex withdraw
+			T::CallOcex::on_withdraw(
+				main_account,
+				proxy_account,
+				asset,
+				amount.saturated_into(),
+				do_force_withdraw,
+			)?;
+			Ok(())
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		pub fn get_pallet_account() -> T::AccountId {
+			T::PalletId::get().into_account_truncating()
+		}
+
+		pub fn generate_proxy_account(
```

Comment:
what's the point of these helpers and pseudo-randomness with downcasts to u8 from u32? is governance so lazy that it can not provide existing accounts?

Reply:
We can't ask them to provide accounts.

Reply:
well then I'd recommend to construct [u8; 32] from (8 x u32.to_le_bytes()) which will guarantee uniqueness and avoid downcasts and "overflowing" bitshifts.

Reply:
Done